Ensure compatibility with numpy 2.0

Numpy 2.0 was just released, and when I install the latest version (`numpy=2.0.0`) into an environment and run `python -m pytest tests`, I'm getting a few failures on my mac:

```
FAILED tests/test_geoprocessing.py::TestGeoprocessing::test_raster_map_nodata - TypeError: can_cast() does not support Python ints, floats, and complex because the result used to depend on the value.
FAILED tests/test_geoprocessing.py::TestGeoprocessing::test_raster_map_nodata_and_dtype - TypeError: can_cast() does not support Python ints, floats, and complex because the result used to depend on the value.
FAILED tests/test_geoprocessing.py::TestGeoprocessing::test_raster_map_nodata_propagates - TypeError: can_cast() does not support Python ints, floats, and complex because the result used to depend on the value.
FAILED tests/test_geoprocessing.py::TestGeoprocessing::test_reclassify_raster_reclass_max_nodata - TypeError: in method 'Band_SetNoDataValue', argument 2 of type 'double'
```

Since this is a major api change, maybe we should expand our test matrix to test against numpy 2.0 if we aren't already?

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -58,7 +58,7 @@ jobs:
               python-build
               flake8
               pytest
-              numpy<2
+              numpy<2  # temporarily capping numpy<2 until we figure out build issues with gdal>=3.8
           environment-name: pyenv
 
       - name: Lint with flake8

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -71,7 +71,7 @@ jobs:
 
       - name: Build and Install PyGeoprocessing
         run: |
-            python -m build --wheel --verbose
+            python -m build --wheel --verbose --no-isolation
             python -m pip install $(find dist -name "*.whl")
 
       - name: Test with pytest

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -71,7 +71,7 @@ jobs:
 
       - name: Build and Install PyGeoprocessing
         run: |
-            python -m build --wheel
+            python -m build --wheel --verbose
             python -m pip install $(find dist -name "*.whl")
 
       - name: Test with pytest

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -58,10 +58,7 @@ jobs:
               python-build
               flake8
               pytest
-              cython>=3
-              setuptools_scm
-              wheel
-              numpy
+              numpy<2
           environment-name: pyenv
 
       - name: Lint with flake8
@@ -75,7 +72,7 @@ jobs:
 
       - name: Build and Install PyGeoprocessing
         run: |
-            python -m build --wheel --verbose --no-isolation
+            python -m build --wheel --verbose
             python -m pip install $(find dist -name "*.whl")
 
       - name: Test with pytest

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -58,6 +58,10 @@ jobs:
               python-build
               flake8
               pytest
+              cython>=3
+              setuptools_scm
+              wheel
+              numpy
           environment-name: pyenv
 
       - name: Lint with flake8

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -51,6 +51,7 @@ jobs:
           condarc: |
               channels:
                 - conda-forge
+          # temporarily capping numpy<2 until we figure out build issues with gdal>=3.8
           create-args: >-
               python=${{ matrix.python-version }}
               gdal=${{ matrix.gdal }}
@@ -58,7 +59,7 @@ jobs:
               python-build
               flake8
               pytest
-              numpy<2  # temporarily capping numpy<2 until we figure out build issues with gdal>=3.8
+              numpy<2
           environment-name: pyenv
 
       - name: Lint with flake8

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -1,6 +1,11 @@
 Release History
 ===============
 
+Unreleased Changes
+------------------
+* Updating for numpy 2.0 API changes.  Pygeoprocessing is now compatible with
+  numpy 2.0 and later.  https://github.com/natcap/pygeoprocessing/issues/396
+
 2.4.4 (2024-05-21)
 ------------------
 * Our github actions for building python distributions now use

--- a/setup.py
+++ b/setup.py
@@ -51,6 +51,7 @@ setup(
                 'src/pygeoprocessing/routing'],
             extra_compile_args=compiler_and_linker_args,
             extra_link_args=compiler_and_linker_args,
+            define_macros=[('NPY_NO_DEPRECATED_API', 'NPY_1_7_API_VERSION')],
             language="c++",
         ),
         Extension(
@@ -61,6 +62,7 @@ setup(
                 'src/pygeoprocessing/routing'],
             extra_compile_args=compiler_and_linker_args,
             extra_link_args=compiler_and_linker_args,
+            define_macros=[('NPY_NO_DEPRECATED_API', 'NPY_1_7_API_VERSION')],
             language="c++",
         ),
         Extension(
@@ -70,6 +72,7 @@ setup(
             include_dirs=[numpy.get_include()],
             extra_compile_args=compiler_and_linker_args,
             extra_link_args=compiler_and_linker_args,
+            define_macros=[('NPY_NO_DEPRECATED_API', 'NPY_1_7_API_VERSION')],
             language="c++"
         ),
     ])

--- a/src/pygeoprocessing/geoprocessing.py
+++ b/src/pygeoprocessing/geoprocessing.py
@@ -696,7 +696,7 @@ def raster_map(op, rasters, target_path, target_nodata=None, target_dtype=None,
     if target_nodata is None:
         target_nodata = choose_nodata(target_dtype)
     else:
-        if not numpy.can_cast(target_nodata, target_dtype):
+        if numpy.result_type(target_nodata, target_dtype) != target_dtype:
             raise ValueError(
                 f'Target nodata value {target_nodata} is incompatible with '
                 f'the target dtype {target_dtype}')

--- a/src/pygeoprocessing/geoprocessing.py
+++ b/src/pygeoprocessing/geoprocessing.py
@@ -4239,6 +4239,12 @@ def numpy_array_to_raster(
             "Origin and pixel size must both be defined or both be None")
     new_band = new_raster.GetRasterBand(1)
     if target_nodata is not None:
+        if numpy.issubdtype(type(target_nodata), numpy.floating):
+            target_nodata = float(target_nodata)
+        elif numpy.issubdtype(type(target_nodata), numpy.integer):
+            target_nodata = int(target_nodata)
+        # Explicitly leaving off an else clause in case there's an edge case we
+        # don't know about.  If so, we should wait for GDAL to raise an error.
         new_band.SetNoDataValue(target_nodata)
     new_band.WriteArray(base_array)
     new_band = None

--- a/src/pygeoprocessing/geoprocessing.py
+++ b/src/pygeoprocessing/geoprocessing.py
@@ -696,7 +696,8 @@ def raster_map(op, rasters, target_path, target_nodata=None, target_dtype=None,
     if target_nodata is None:
         target_nodata = choose_nodata(target_dtype)
     else:
-        if numpy.result_type(target_nodata, target_dtype) != target_dtype:
+        if not numpy.can_cast(numpy.min_scalar_type(target_nodata),
+                              target_dtype):
             raise ValueError(
                 f'Target nodata value {target_nodata} is incompatible with '
                 f'the target dtype {target_dtype}')

--- a/src/pygeoprocessing/geoprocessing_core.pyx
+++ b/src/pygeoprocessing/geoprocessing_core.pyx
@@ -13,6 +13,7 @@ cimport cython
 cimport libc.math as cmath
 cimport libcpp.algorithm
 cimport numpy
+numpy.import_array()
 from cython.operator cimport dereference as deref
 from cython.operator cimport preincrement as inc
 from libc.stdio cimport fclose

--- a/src/pygeoprocessing/routing/routing.pyx
+++ b/src/pygeoprocessing/routing/routing.pyx
@@ -38,6 +38,7 @@ import time
 
 cimport cython
 cimport numpy
+numpy.import_array()
 from cpython.mem cimport PyMem_Malloc, PyMem_Free
 from cython.operator cimport dereference as deref
 from cython.operator cimport preincrement as inc

--- a/src/pygeoprocessing/routing/watershed.pyx
+++ b/src/pygeoprocessing/routing/watershed.pyx
@@ -6,6 +6,7 @@ import time
 
 cimport cython
 cimport numpy
+numpy.import_array()
 from cpython.mem cimport PyMem_Malloc, PyMem_Free
 from cython.operator cimport dereference as deref
 from cython.operator cimport preincrement as inc

--- a/tests/test_geoprocessing.py
+++ b/tests/test_geoprocessing.py
@@ -309,8 +309,8 @@ class TestGeoprocessing(unittest.TestCase):
         test_value = 0.5
         pixel_matrix[:] = test_value
         nodata = numpy.finfo(numpy.float32).max - 1
-        pixel_matrix[0,0] = nodata
-        pixel_matrix[5,7] = nodata
+        pixel_matrix[0, 0] = nodata
+        pixel_matrix[5, 7] = nodata
         raster_path = os.path.join(self.workspace_dir, 'raster.tif')
         target_path = os.path.join(self.workspace_dir, 'target.tif')
         _array_to_raster(

